### PR TITLE
Adding functional tests.

### PR DESCRIPTION
**Pull request by jdalton**
Start pushing functional tests over.

### CODE REVIEW

File: tests/functional/pointerevent_constructor.js [hunk 1]
```diff
@@ -1,16 +1,13 @@
 define(function(require) {
 	var registerSuite = require('intern!object');
 	var w3cTest = require('../support/w3cTest');
-	var name = 'pointerevent_button_attribute_mouse-manual';
+	var name = 'pointerevent_constructor';
```
Comment (scottgonzalez) [inline tests/functional/pointerevent_constructor.js]: Why did you remove the existing test?

File: tests/functional/pointerevent_capture_mouse-manual.js [hunk 2]
```diff
@@ -9,22 +9,22 @@ define(function(require) {
 		main: function() {
 			return w3cTest(this.remote, name + '.html')
 				.findById('target0')
-					.moveMouseTo()
-				.end()
+					.moveMouseTo(50, 25)
+					.clickMouseButton(0)
```
Comment (scottgonzalez) [inline tests/functional/pointerevent_capture_mouse-manual.js]: Why were all of these clicks added?

File: tests/functional/pointerevent_capture_mouse-manual.js [hunk 3]
```diff
@@ -9,22 +9,22 @@ define(function(require) {
...
+					.end()
```
Comment (scottgonzalez) [inline tests/functional/pointerevent_capture_mouse-manual.js]: Why did the all the `.end()` calls get indented?

File: tests/functional/pointerevent_capture_suppressing_mouse-manual.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+					.end()
+				.findById('target1')
+					.moveMouseTo(50, 30)
+					.clickMouseButton(0)
+					.end()
+				.findById('btnCapture')
+					.moveMouseTo(50, 4)
+					.pressMouseButton(0)
+					.end()
+				.findByTagName('body')
```
Comment (scottgonzalez) [inline tests/functional/pointerevent_capture_suppressing_mouse-manual.js]: This is missing three steps:

* Put your mouse over the purple rectangle and then move it out. Nothing should happen
* Put your mouse over the black rectangle. pointerover and pointerenter should be logged inside of it.
* Move your mouse out of the black rectangle. pointerout and pointerleave should be logged inside of it

File: tests/functional/pointerevent_constructor.js (hunk 1 above)
Comment (jdalton) [inline tests/functional/pointerevent_constructor.js]: This looks like some git merge fun. Will check it out.

File: tests/functional/pointerevent_capture_suppressing_mouse-manual.js (hunk 4 above)
Comment (jdalton) [inline tests/functional/pointerevent_capture_suppressing_mouse-manual.js]: I'll double check but if it's passing the test then it's passing the test.

File: tests/functional/pointerevent_capture_mouse-manual.js (hunk 3 above)
Comment (jdalton) [inline tests/functional/pointerevent_capture_mouse-manual.js]: Formatting nit.

File: tests/functional/pointerevent_capture_mouse-manual.js (hunk 2 above)
Comment (jdalton) [inline tests/functional/pointerevent_capture_mouse-manual.js]: Most likely in the course of recording the actions.

**Comment by jdalton** [comment]
Separate PRs isn't going to work for this as tests end up requiring work from other PRs. I'll flatten these 2 down.

File: tests/functional/pointerevent_capture_suppressing_mouse-manual.js (hunk 4 above)
Comment (scottgonzalez) [inline tests/functional/pointerevent_capture_suppressing_mouse-manual.js]: Well, the assertion is that you don't get events fired on the targets because capture is enabled. If you don't hover over the targets, then of course the events won't fire. It's possible that the test isn't defensive enough to ensure that all the necessary steps are taken to prevent false positives.

Comment (jdalton) [inline tests/functional/pointerevent_capture_suppressing_mouse-manual.js]: Ok checked. These fall out naturally when moving the mouse from the `btnCapture` to the `body` coordinates. It's basically going down the page then back up. Without moving the left coordinate.

File: tests/functional/pointerevent_pointerleave_descendants-manual.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+	registerSuite({
+		name: name,
+
+		main: function() {
+			return w3cTest(this.remote, name + '.html')
+				.findById('target0')
+					.moveMouseTo(50, 8)
+					.end()
+				.findByTagName('body')
+					.moveMouseTo(50, 30)
```
Comment (scottgonzalez) [inline tests/functional/pointerevent_pointerleave_descendants-manual.js]: This seems like it would move the cursor up from the black box to the body, not down through the purple box.

Comment (jdalton) [inline tests/functional/pointerevent_pointerleave_descendants-manual.js]: I believe the position is sufficient enough to get the desired effect (mouse move downing) to cause the test to complete.

Comment (scottgonzalez) [inline tests/functional/pointerevent_pointerleave_descendants-manual.js]: 30px from the top of the body should be above the black box, not below the purple box. I'm not sure how this would work. Am I not understanding how the values work for `.moveMouseTo()`?

Comment (jdalton) [inline tests/functional/pointerevent_pointerleave_descendants-manual.js]: Ya, I've tried different combinations without any luck (the test would fail).

Maybe @csnover has some insight into the `moveMouseTo` positioning and element anchoring bits.

Comment (jdalton) [inline tests/functional/pointerevent_pointerleave_descendants-manual.js]: Ping @csnover :D

Comment (csnover) [inline tests/functional/pointerevent_pointerleave_descendants-manual.js]: Sorry, I am A) still a hot mess and B) get too many notifications from GitHub.

The coordinates in `moveMouseTo` are relative to the top-left corner of the context element (the thing you got using `find*`). So as @scottgonzalez notes this test starts the mouse at 50,8 from the corner of `#target0` and then moves it to 50,30 of `body`, which moves the mouse upwards and never contacts the purple box.

Looking at the test instructions, I suspect a key thing to understand given the direction to “move it out through purple rectangle boundaries” is that you will never get intermediate `mousemove/over/out` events like you would from a hardware mouse device. WebDriver will move the mouse from being in one place in the first tick to being exactly in the requested next place the next tick. If there are elements in between these two points, they won’t be triggered. So while you might intuitively think “oh I can just start the mouse at the top of the document and move it to the bottom of the document and it will move through all the intersecting elements”, in reality it will just always be in the body. You’ll need to add those intermediate positions yourself explicitly.

Hopefully that helps, let me know if I misunderstand something about the question.

Comment (jdalton) [inline tests/functional/pointerevent_pointerleave_descendants-manual.js]: Tested it being explicit with the actions and it still passes so \o/
